src/check/canonicalize/Scope.zig

```diff
@@ -52,26 +62,99 @@ pub const IntroduceResult = union(enum) {
     var_across_function_boundary: CIR.Pattern.Idx,
 };
 
+/// Result of introducing a type declaration
+pub const TypeIntroduceResult = union(enum) {
+    success: void,
+    shadowing_warning: CIR.Statement.Idx, // The type declaration that was shadowed
+    redeclared_error: CIR.Statement.Idx, // The type declaration that was redeclared
+    type_alias_redeclared: CIR.Statement.Idx, // The type alias that was redeclared
+    custom_type_redeclared: CIR.Statement.Idx, // The custom type that was redeclared
+    cross_scope_shadowing: CIR.Statement.Idx, // Type shadowed across different scopes
+    parameter_conflict: struct {
+        original_stmt: CIR.Statement.Idx,
+        conflicting_parameter: base.Ident.Idx,
+    },
+};
+
 /// Item kinds in a scope
-pub const ItemKind = enum { ident, alias };
+pub const ItemKind = enum { ident, alias, type_decl };
 
 /// Get the appropriate map for the given item kind
-pub fn items(scope: *Scope, comptime item_kind: ItemKind) *std.AutoHashMapUnmanaged(Ident.Idx, CIR.Pattern.Idx) {
+pub fn items(scope: *Scope, comptime item_kind: ItemKind) switch (item_kind) {
+    .ident, .alias => *std.AutoHashMapUnmanaged(Ident.Idx, CIR.Pattern.Idx),
+    .type_decl => *std.AutoHashMapUnmanaged(Ident.Idx, CIR.Statement.Idx),
+} {
     return switch (item_kind) {
         .ident => &scope.idents,
         .alias => &scope.aliases,
+        .type_decl => &scope.type_decls,
     };
 }
 
 /// Get the appropriate map for the given item kind (const version)
-pub fn itemsConst(scope: *const Scope, comptime item_kind: ItemKind) *const std.AutoHashMapUnmanaged(Ident.Idx, CIR.Pattern.Idx) {
+pub fn itemsConst(scope: *const Scope, comptime item_kind: ItemKind) switch (item_kind) {
+    .ident, .alias => *const std.AutoHashMapUnmanaged(Ident.Idx, CIR.Pattern.Idx),
+    .type_decl => *const std.AutoHashMapUnmanaged(Ident.Idx, CIR.Statement.Idx),
+} {
     return switch (item_kind) {
         .ident => &scope.idents,
         .alias => &scope.aliases,
+        .type_decl => &scope.type_decls,
     };
 }
 
 /// Put an item in the scope, panics on OOM
-pub fn put(scope: *Scope, gpa: std.mem.Allocator, comptime item_kind: ItemKind, name: Ident.Idx, pattern: CIR.Pattern.Idx) void {
-    scope.items(item_kind).put(gpa, name, pattern) catch |err| collections.utils.exitOnOom(err);
+pub fn put(scope: *Scope, gpa: std.mem.Allocator, comptime item_kind: ItemKind, name: Ident.Idx, value: switch (item_kind) {
+    .ident, .alias => CIR.Pattern.Idx,
+    .type_decl => CIR.Statement.Idx,
+}) void {
+    scope.items(item_kind).put(gpa, name, value) catch |err| collections.utils.exitOnOom(err);
+}
+
+/// Introduce a type declaration into the scope
+pub fn introduceTypeDecl(
+    scope: *Scope,
+    gpa: std.mem.Allocator,
+    ident_store: *const base.Ident.Store,
+    name: Ident.Idx,
+    type_decl: CIR.Statement.Idx,
+    parent_lookup_fn: ?fn (Ident.Idx) ?CIR.Statement.Idx,
+) TypeIntroduceResult {
```

Comment:
TODO look closer at this implementation... I'm not sure we want to be handling type declarations like this in Scope

Reply:
An important thing we want to preserve is that if this is an annotation for an assignment, then all type annotations in that assignment expression should share named idents with this annotation - see [my earlier comment](https://github.com/roc-lang/roc/pull/7849#discussion_r2160089296) for an example!